plugins/azure_sentinel/icon_azure_sentinel/connection/connection.py

```diff
@@ -0,0 +1,89 @@
+import time
+
+import insightconnect_plugin_runtime
+import requests
+from icon_azure_sentinel.util.api import AzureSentinelClient
+from insightconnect_plugin_runtime.exceptions import ConnectionTestException, PluginException
+
+from .schema import ConnectionSchema, Input
+
+# Custom imports below
+
+
+class Connection(insightconnect_plugin_runtime.Connection):
+    def __init__(self):
+        super(self.__class__, self).__init__(input=ConnectionSchema())
+        self.O365_AUTH_ENDPOINT = "https://login.microsoftonline.com/{}/oauth2/token"
+        self.SCOPE = "https://management.azure.com"
+        self.tenant = ""
+        self.client_secret = ""  # nosec
+        self.time_ago = 0  # Jan 1, 1970
+        self.time_now = time.time()  # More than 1 hour since 1978
+        self.client_id = ""
+        self.auth_token = ""  # nosec
+        self.api_client = None
+
+    def connect(self, params):
+        self.logger.info("Connect: Connecting...")
+        self.tenant = params.get(Input.TENANT_ID)
+        self.client_id = params.get(Input.CLIENT_ID)
+        self.client_secret = params.get(Input.CLIENT_SECRET).get("secretKey")
+        self.auth_token = self.get_auth_token()
+        self.api_client = AzureSentinelClient(self.auth_token)
+
+    def get_auth_token(self):
+        tenant_id = self.tenant
+        client_id = self.client_id
+        client_secret = self.client_secret
+
+        self.time_now = time.time()
+        if (self.time_now - self.time_ago) < 3500:  # 1 hour in seconds
+            return self.auth_token
+
+        self.logger.info("Updating auth token...")
+
+        data = {
+            "grant_type": "client_credentials",
+            "client_id": client_id,
+            "resource": self.SCOPE,
+            "client_secret": client_secret,
+        }
+
+        formatted_endpoint = self.O365_AUTH_ENDPOINT.format(tenant_id)
+        self.logger.info("Getting token from: " + formatted_endpoint)
+
+        request = requests.request("POST", formatted_endpoint, data=data)
+        self.logger.info("Authentication request status: " + str(request.status_code))
+
+        if request.status_code != 200:
+            self.logger.error(request.text)
+            raise PluginException(
+                cause="Unable to authorize against Microsoft graph API.",
+                assistance="The application may not be authorized to connect "
+                "to the Microsoft Graph API. Please contact your "
+                "Azure administrator.",
+                data=request.text,
+            )
+
+        token = request.json().get("access_token")
+
+        self.time_ago = time.time()
+        self.auth_token = token
+
+        self.logger.info(f"Authentication Token: ****************{self.auth_token[-5:]}")
+
+        return token
+
+    @staticmethod
+    def get_headers(auth_token):
+        headers = {"Content-type": "application/json", "Authorization": "Bearer " + auth_token}
+        return headers
+
```

Comment:
Would move those functions into AzureSentinelClient class, but it's up to you :smiley:

Reply:
Done!